sound/soc/sof/sof-client.c

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/debugfs.h>
+#include <linux/errno.h>
+#include <linux/list.h>
+#include <linux/mutex.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include "sof-client.h"
+#include "sof-priv.h"
+
+static void sof_client_connect(struct ancillary_device *adev)
+{
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+	struct snd_sof_dev *sdev = cdev->sdev;
+
+	/* add to list of SOF client devices */
+	mutex_lock(&sdev->client_mutex);
+	list_add(&cdev->list, &sdev->client_list);
+	mutex_unlock(&sdev->client_mutex);
+}
+
+static void sof_client_disconnect(struct ancillary_device *adev)
+{
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+	struct snd_sof_dev *sdev = cdev->sdev;
+
+	/* remove from list of SOF client devices */
+	mutex_lock(&sdev->client_mutex);
+	list_del(&cdev->list);
+	mutex_unlock(&sdev->client_mutex);
+}
+
+static void sof_client_adev_release(struct device *dev)
+{
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+
+	ida_simple_remove(cdev->client_ida, adev->id);
+	kfree(cdev);
+}
+
+int sof_client_dev_register(struct snd_sof_dev *sdev, const char *name, struct ida *client_ida)
+{
+	struct sof_client_dev *cdev;
+	struct ancillary_device *adev;
+	int ret;
+
+	cdev = kzalloc(sizeof(*cdev), GFP_KERNEL);
+	if (!cdev)
+		return -ENOMEM;
+
+	cdev->sdev = sdev;
+	cdev->connect = sof_client_connect;
+	cdev->disconnect = sof_client_disconnect;
+	cdev->client_ida = client_ida;
+	adev = &cdev->adev;
+	adev->name = name;
+	adev->dev.parent = sdev->dev;
+	adev->dev.release = sof_client_adev_release;
+
+	adev->id = ida_alloc(client_ida, GFP_KERNEL);
+	if (adev->id < 0) {
+		dev_err(sdev->dev, "error: get IDA idx for ancillary device %s failed\n", name);
+		kfree(cdev);
+		return adev->id;
+	}
+
+	ret = ancillary_device_register(adev);
+	if (ret < 0) {
+		kfree(cdev);
```

Comment:
that's broken, you can't identify a case where the register fails in the first three tests or after device_initialize.

This makes me wonder if you need to split ancillary_device_register in two, one for initialization and one for device_add.

Reply:
I am going to rewrite register to always go through put_device on error state and return an error value.  In unregister, I am going to do:
       `if (adev->match_name)
                   kfree(adev->match_name)`
to avoid double freeing match_name.

the parent can always assume that put_device will be called on a failed registration.  If they don't provide a release function then it will fail to register and they will get an error message in the log why.   If they provide a release callback, then they can rely on that to free the memory.  Up to them how they want to code it, but not relevant to ancillary_bus.